src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,407 @@
+    return some val
+  return none
+
+structure IntroDefResult where
+  -- the new name/fvar of the defn.
+  defn : FVarId
+  -- a proof 'hdefn : name = value
+  eqProof : FVarId
+
+/-
+Introduce a new definition with name `name : hdefTy` into the local context,
+and return the FVarId of the new definition in the new goal (the MVarId) returned.
+-/
+def introDefExt (g : MVarId) (name : Name) (hdefTy : Expr) (hdefVal : Expr) : AckM (IntroDefResult × MVarId) := do
+  withContext g do
+    let g ← g.assertExt name (hName := Name.str name "value") hdefTy hdefVal
+    let (defn, g) ← g.intro1P
+    let (eqProof, g) ← g.intro1P
+    return ({ defn, eqProof}, g)
+
+/-- Insert the CallVal `cv` at `(fn, args)` into the state. -/
+private def _insertCallVal (fn : Function) (args : ArgumentList) (cv : CallVal) : AckM Unit := do
+  let calls ← getCallMap fn
+  modify fun s => { s with fn2args2call := s.fn2args2call.insert fn (calls.insert args cv) }
+
+/--
+Replace a call to the function `f` with an `fvar`. Since the `fvar` is defeq to the call,
+we can just replace occurrences of the call with the fvar `f`.
+
+We will later need to add another hypothesis with the equality that `fvar = f x₁ ... xₙ`
+-/
+def replaceCallWithFVar (g : MVarId) (fn : Function) (args : ArgumentList) : AckM (CallVal × MVarId) := do
+  g.withContext do
+    if let some val ← getCallVal? fn args then
+      trace[bv_ack] "using cached call {val} for {fn} {args}"
+      return (val, g)
+    let e := mkAppN fn.f (args.map Argument.x)
+    let name ← gensym
+    let (introDef, g) ← introDefExt g name fn.codomain.toExpr e
+    let cv := { fvar := introDef.defn, heqProof := Expr.fvar introDef.eqProof : CallVal }
+    _insertCallVal fn args cv
+    return (cv, g)
+
+/-- Create a trace node in trace class (i.e. `set_option traceClass true`),
+with header `header`, whose default collapsed state is `collapsed`. -/
+def withTraceNode (header : MessageData) (k : AckM α)
+    (collapsed : Bool := true)
+    (traceClass : Name := `bv_ack) : AckM α :=
+  Lean.withTraceNode traceClass (fun _ => return header) k (collapsed := collapsed)
+
+/-- Returns `True` if the type is a function type that is understood by the bitblaster. -/
+def isBitblastTy (e : Expr) : Bool :=
+  match_expr e with
+  | BitVec _ => true
+  | Bool => true
+  | _ => false
+
+mutual
+partial def ackAppChildren (g : MVarId) (e : Expr) : AckM (Expr × MVarId) := do
+  g.withContext do
+    trace[bv_ack] "{crossEmoji} bailing out on illegal application {e}"
+    let f := e.getAppFn
+    let (f, g) ← introAckForExpr g f
+    -- NOTE: Rewrite as a fold? It's not too much cleaner, sadly.
+    let mut args := #[]
+    let mut g := g
+    for arg in e.getAppArgs do
+      let gArg ← introAckForExpr g arg
+      g := gArg.2
+      args := args.push gArg.1
+    return (mkAppN f args, g) -- NOTE: is there some way to use update to update this?
+
+
+partial def introAckForExpr (g : MVarId) (e : Expr) : AckM (Expr × MVarId) := do
```

Comment:
Going directly under binders violates the invariant that we don't deal with loose bvars in `MetaM`, all important functions like `isDefEq`, `whnf` etc. rely on the fact that you do not do this. Is there a particular reason that you break the invariant here.

Furthermore despite being some of the larger functions in this file these are one of the only ones without a doc string.